collectors/state_collector.go

```diff
@@ -0,0 +1,105 @@
+package collectors
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/lightninglabs/lndclient"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type StateCollector struct {
+	lnd *lndclient.LndServices
+
+	// Use one gauge to track the starting time of LND.
+	timeToStartDesc *prometheus.Desc
+
+	// startTime records a best-effort timestamp of when LND was started.
+	startTime time.Time
+
+	// endTime records when LND makes a transition from RPC_ACTIVE to
+	// SERVER_ACTIVE.
+	// If lndmon starts after LND has already reached SERVER_ACTIVE, no
+	// startup time metric will be emitted.
+	endTime time.Time
+
+	// mutex is a lock for preventing concurrent writes to startTime or
+	// endTime.
+	mutex sync.RWMutex
+
+	// errChan is a channel that we send any errors that we encounter into.
+	// This channel should be buffered so that it does not block sends.
+	errChan chan<- error
+}
+
+// NewStateCollector returns a new instance of the StateCollector.
+func NewStateCollector(lnd *lndclient.LndServices,
+	errChan chan<- error) *StateCollector {
+
+	sc := &StateCollector{
+		lnd: lnd,
+		timeToStartDesc: prometheus.NewDesc(
+			"lnd_time_to_start_secs",
+			"time to start in seconds",
+			nil, nil,
+		),
+		startTime: time.Now(),
+		errChan:   errChan,
+	}
+
+	go sc.monitorStateChanges()
+	return sc
+}
+
+// monitorStateChanges checks the state every second to catch fast transitions.
+func (s *StateCollector) monitorStateChanges() {
+	var serverActiveReached bool
+
+	for {
+		state, err := s.lnd.State.GetState(context.Background())
+		if err != nil {
+			s.errChan <- fmt.Errorf("StateCollector GetState failed with: %v", err)
+			continue
+		}
+
+		s.mutex.Lock()
+		if state == lndclient.WalletStateRPCActive && !s.startTime.IsZero() {
+			s.endTime = time.Now()
+			serverActiveReached = true
+		}
+		s.mutex.Unlock()
+
+		if serverActiveReached {
+			break
+		}
+		time.Sleep(1 * time.Second)
```

Comment:
So we can make the entire thing event driven, using the `SubscribeState` call, which should also make the measurements that much more accurate: https://lightning.engineering/api-docs/api/lnd/state/subscribe-state

Reply:
I refactored this to use `SubscribeState` and it breaks when `lndmon` is started prior to LND. Will try and get things working again.